boringtun/src/device/mod.rs

```diff
@@ -290,44 +290,91 @@ impl Device {
     fn update_peer(
         &mut self,
         pub_key: X25519PublicKey,
+        update_only: bool,
         remove: bool,
-        _replace_ips: bool,
+        replace_ips: bool,
         endpoint: Option<SocketAddr>,
         allowed_ips: &[AllowedIP],
         keepalive: Option<u16>,
         preshared_key: Option<[u8; 32]>,
-    ) {
+    ) -> Result<(), Error> {
         let pub_key = Arc::new(pub_key);
+
         if remove {
-            // Completely remove a peer
-            return self.remove_peer(&pub_key);
+            self.remove_peer(&pub_key);
+
+            return Ok(());
         }
 
-        // Update an existing peer
-        if self.peers.get(&pub_key).is_some() {
-            // We already have a peer, we need to merge the existing config into the newly created one
-            panic!("Modifying existing peers is not yet supported. Remove and add again instead.");
+        if let Some(peer) = self.peers.get(&pub_key) {
+            let peer = Arc::clone(peer);
+
+            if let Some(endpoint) = endpoint {
+                peer.set_endpoint(endpoint);
+            }
+
+            if replace_ips {
+                self.peers_by_ip
+                    .remove(&|p: &Arc<Peer>| Arc::ptr_eq(&peer, p));
+                peer.set_allowed_ips(allowed_ips);
+            } else {
+                peer.add_allowed_ips(allowed_ips);
+            }
+
+            if let Some(keepalive) = keepalive {
+                peer.set_persistent_keepalive(keepalive);
+            }
+
+            if let Some(preshared_key) = preshared_key {
+                peer.set_preshared_key(preshared_key);
+            }
+
+            for AllowedIP { addr, cidr } in allowed_ips {
+                self.peers_by_ip
+                    .insert(*addr, *cidr as _, Arc::clone(&peer));
+            }
+        } else {
+            if update_only {
+                return Ok(());
+            }
+
+            return self
+                .new_peer(pub_key, endpoint, allowed_ips, keepalive, preshared_key)
+                .and(Ok(()));
         }
 
+        Ok(())
+    }
+
+    fn new_peer(
+        &mut self,
+        pub_key: Arc<X25519PublicKey>,
+        endpoint: Option<SocketAddr>,
+        allowed_ips: &[AllowedIP],
+        keepalive: Option<u16>,
+        preshared_key: Option<[u8; 32]>,
+    ) -> Result<Arc<Peer>, Error> {
         let next_index = self.next_index();
         let device_key_pair = self
             .key_pair
             .as_ref()
             .expect("Private key must be set first");
-
         let tunn = Tunn::new(
             Arc::clone(&device_key_pair.0),
             Arc::clone(&pub_key),
             preshared_key,
             keepalive,
             next_index,
             None,
-        )
-        .unwrap();
-
-        let peer = Peer::new(tunn, next_index, endpoint, &allowed_ips, preshared_key);
+        );
+        let peer = Arc::new(Peer::new(
+            tunn.unwrap(),
```

Comment:
Since you already return a `Result` in this function, you could get rid of the `.unwrap()` call.